src/api/accounts.rs

```diff
@@ -0,0 +1,515 @@
+use crate::Transport;
+use ethsign::{Error as EthsignError, SecretKey, Signature};
+use futures::future::{self, Either, FutureResult, Join3};
+use futures::{Async, Future, Poll};
+use parity_crypto::Keccak256;
+use rlp::RlpStream;
+
+/// Transaction base fee, this will be used as the default transaction fee in
+/// the case none is specified for signing.
+pub const TRANSACTION_BASE_GAS_FEE: U256 = U256([21000, 0, 0, 0]);
+
+/// `Accounts` namespace
+#[derive(Debug, Clone)]
+pub struct Accounts<T> {
+    transport: T,
+}
+
+impl<T: Transport> Namespace<T> for Accounts<T> {
+    fn new(transport: T) -> Self
+    where
+        Self: Sized,
+    {
+        Accounts { transport }
+    }
+
+    fn transport(&self) -> &T {
+        &self.transport
+    }
+}
+
+impl<T: Transport> Accounts<T> {
+    /// Gets the parent `web3` namespace
+    fn web3(&self) -> Web3<T> {
+        Web3::new(self.transport.clone())
+    }
+
+    /// Signs an Ethereum transaction with a given private key.
+    pub fn sign_transaction(&self, tx: TransactionParameters, key: &SecretKey) -> SignTransactionFuture<T> {
+        SignTransactionFuture::new(self, tx, key)
+    }
+
+    /// Hash a message. The data will be UTF-8 HEX decoded and enveloped as
+    /// follows: `"\u{19}Ethereum Signed Message:\n" + message.length + message`
+    /// and hashed using keccak256.
+    pub fn hash_message<S>(&self, message: S) -> H256
+    where
+        S: AsRef<str>,
+    {
+        let message = message.as_ref();
+        let eth_message = format!("\u{0019}Ethereum Signed Message:\n{}{}", message.len(), message);
+
+        eth_message.as_bytes().keccak256().into()
+    }
+
+    /// Sign arbitrary string data. The data is UTF-8 encoded and enveloped the
+    /// same way as with `hash_message`.
+    pub fn sign<S>(&self, message: S, key: &SecretKey) -> Result<SignedData, Error>
+    where
+        S: AsRef<str>,
+    {
+        let message = message.as_ref().to_string();
+        let message_hash = self.hash_message(&message);
+
+        let signature = key.sign(&message_hash[..]).map_err(EthsignError::from)?;
+        // this is what web3.js does ¯\_(ツ)_/¯, it is documented here:
+        // https://web3js.readthedocs.io/en/v1.2.2/web3-eth-accounts.html#sign
+        let v = signature.v + 27;
+
+        let signature_bytes = Bytes({
+            let mut bytes = Vec::with_capacity(65);
+            bytes.extend_from_slice(&signature.r[..]);
+            bytes.extend_from_slice(&signature.s[..]);
+            bytes.push(v);
+            bytes
+        });
+
+        Ok(SignedData {
+            message,
+            message_hash: message_hash.into(),
+            v,
+            r: signature.r.into(),
+            s: signature.s.into(),
+            signature: signature_bytes,
+        })
+    }
+
+    /// Recovers the Ethereum address which was used to sign the given data.
+    pub fn recover<S, Sig>(&self, message: S, signature: Sig) -> Result<Address, Error>
```

Comment:
Might be useful to have another `recover` for raw hashes? Maybe make the `message` an `enum` and allow passing etiher `Hash` or `AsRef<str>`?

Reply:
Oops! That was a bit of an oversight.

Reply:
@tomusdrw Actually, thinking about it now it would be nice to be able to do something like:
```rust
let signed_data = accounts.sign("hello world", &key);
let address = accounts.recover(signed_data)?;
assert_eq!(address, key.public().address());
let signed_tx = accounts.sign(tx, &key).await?;
let address = accounts.recover(signed_tx)?;
assert_eq!(address, key.public().address());
```

What do you think?

Reply:
Yes, but how do you recover something you didn't sign yourself? Also since we don't have specialisation either the method names need to be different or you need to accept `impl Into<SomeEnum>` as a parameter.

Reply:
I went ahead and implemented this. If you don't like it, let me know and I can change it.

Reply:
Sorry, for some reason I didn't see you message when I commented. Yes you can recover things that you did not sign yourself. A `Recovery` struct with the recovery data can be constructed fairly easily, we do this in one of the unit tests.

Let me know if you like how I implemented it, if not I can change it to something else.